File: sesion5-lab/spring-security-audit/audit-server/src/main/java/com/example/audit/web/AuditController.java
package com.example.audit.web;

import java.util.LinkedHashMap;
import java.util.Map;

import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.AuthenticationException;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.example.audit.dto.LoginRequest;
import com.example.audit.audit.SecurityAuditListener;
import com.example.audit.service.VulnerableAuditService;
import com.example.audit.support.InMemoryLogAppender;

import jakarta.servlet.http.HttpServletRequest;

@RestController
@RequestMapping("/api")
public class AuditController {

    private final VulnerableAuditService vulnerableAudit;
    private final AuthenticationManager authenticationManager;
    private final SecurityAuditListener securityAuditListener;

    public AuditController(
            VulnerableAuditService vulnerableAudit,
            AuthenticationManager authenticationManager,
            SecurityAuditListener securityAuditListener) {
        this.vulnerableAudit = vulnerableAudit;
        this.authenticationManager = authenticationManager;
        this.securityAuditListener = securityAuditListener;
    }

    @PostMapping("/auth/vulnerable/login")
    public ResponseEntity<Map<String, Object>> loginVulnerable(@RequestBody LoginRequest req) {
        InMemoryLogAppender.clear();
        boolean ok = vulnerableAudit.login(req.username(), req.password());
        return ResponseEntity.ok(body(
                "VULNERABLE (texto libre, password en log, sin correlation ID)",
                ok ? "LOGIN_OK" : "LOGIN_FAIL",
                null));
    }

    @PostMapping("/auth/seguro/login")
    public ResponseEntity<Map<String, Object>> loginSeguro(
            @RequestBody LoginRequest req,
            HttpServletRequest request) {

        InMemoryLogAppender.clear();
        long start = System.currentTimeMillis();
        String result;
        int status;

        try {
            authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(req.username(), req.password()));
            result = "LOGIN_OK";
            status = 200;
        } catch (AuthenticationException ex) {
            result = "LOGIN_FAIL";
            status = 401;
        }

        securityAuditListener.logAccess(
                req.username(),
                "/api/auth/seguro/login",
                "POST",
                status,
                System.currentTimeMillis() - start,
                request.getRemoteAddr(),
                request.getHeader("User-Agent"));

        return ResponseEntity.status(status).body(body(
                "SEGURO (JSON estructurado SECURITY_AUDIT + correlation ID)",
                result,
                request.getHeader("X-Correlation-ID")));
    }

    @GetMapping("/orders/vulnerable/{id}")
    public ResponseEntity<Map<String, Object>> orderVulnerable(@PathVariable String id) {
        InMemoryLogAppender.clear();
        vulnerableAudit.logAccess("anonymous", "/api/orders/vulnerable/" + id, 200);
        return ResponseEntity.ok(body(
                "VULNERABLE (log de acceso sin contexto)",
                Map.of("orderId", id, "total", 149.99),
                null));
    }

    @GetMapping("/orders/seguro/{id}")
    public ResponseEntity<Map<String, Object>> orderSeguro(
            @PathVariable String id,
            HttpServletRequest request) {

        InMemoryLogAppender.clear();
        long start = System.currentTimeMillis();
        securityAuditListener.logAccess(
                request.getHeader("X-User-Id") != null ? request.getHeader("X-User-Id") : "anonymous",
                "/api/orders/seguro/" + id,
                "GET",
                200,
                System.currentTimeMillis() - start,
                request.getRemoteAddr(),
                request.getHeader("User-Agent"));

        return ResponseEntity.ok(body(
                "SEGURO (evento ACCESS con metodo, duracion, IP)",
                Map.of("orderId", id, "total", 149.99),
                request.getHeader("X-Correlation-ID")));
    }

    private static Map<String, Object> body(String modo, Object resultado, String correlationId) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("modo", modo);
        response.put("resultado", resultado);
        response.put("correlationId", correlationId);
        response.put("lineasLog", InMemoryLogAppender.snapshot());
        return response;
    }
}

File: sesion5-lab/aspnet-security-audit/audit-demo/Controllers/AuditController.cs
using System.Text.Json;
using AuditDemo.Models;
using AuditDemo.Services;
using AuditDemo.Support;
using Microsoft.AspNetCore.Mvc;

namespace AuditDemo.Controllers;

[ApiController]
[Route("api")]
public class AuditController : ControllerBase
{
    private readonly VulnerableAuditService _vulnerable;
    private readonly LogCapture _logCapture;

    public AuditController(VulnerableAuditService vulnerable, LogCapture logCapture)
    {
        _vulnerable = vulnerable;
        _logCapture = logCapture;
    }

    [HttpPost("auth/vulnerable/login")]
    public IActionResult LoginVulnerable([FromBody] LoginRequest request)
    {
        _logCapture.Clear();
        var ok = _vulnerable.Login(request.Username, request.Password);
        return Ok(Body("VULNERABLE (texto libre, password en log)", ok ? "LOGIN_OK" : "LOGIN_FAIL", null));
    }

    [HttpPost("seguro/auth/login")]
    public IActionResult LoginSeguro([FromBody] LoginRequest request)
    {
        _logCapture.Clear();
        var ok = request.Username == "ana" && request.Password == "Secr3t!";
        var correlationId = HttpContext.Items["CorrelationId"]?.ToString();

        var audit = new
        {
            timestamp = DateTime.UtcNow.ToString("o"),
            level = ok ? "INFO" : "WARN",
            event_type = ok ? "AUTH_SUCCESS" : "AUTH_FAILURE",
            correlation_id = correlationId,
            user_id = request.Username,
            reason = ok ? "LOGIN_OK" : "INVALID_CREDENTIALS"
        };

        _logCapture.Add(JsonSerializer.Serialize(audit));

        return ok
            ? Ok(Body("SEGURO (JSON estructurado + correlation ID)", "LOGIN_OK", correlationId))
            : Unauthorized(Body("SEGURO (JSON estructurado + correlation ID)", "LOGIN_FAIL", correlationId));
    }

    [HttpGet("orders/vulnerable/{id}")]
    public IActionResult OrderVulnerable(string id)
    {
        _logCapture.Clear();
        _vulnerable.LogAccess("anonymous", $"/api/orders/vulnerable/{id}", 200);
        return Ok(Body("VULNERABLE", new { orderId = id, total = 149.99m }, null));
    }

    [HttpGet("seguro/orders/{id}")]
    public IActionResult OrderSeguro(string id)
    {
        _logCapture.Clear();
        return Ok(Body(
            "SEGURO (SecurityAuditMiddleware registra ACCESS)",
            new { orderId = id, total = 149.99m },
            HttpContext.Items["CorrelationId"]?.ToString()));
    }

    private object Body(string modo, object resultado, string? correlationId) => new
    {
        modo,
        resultado,
        correlationId,
        lineasLog = _logCapture.Snapshot()
    };
}
